crates/bevy_pbr/src/material.rs

```diff
@@ -210,11 +209,27 @@ impl<M: SpecializedMaterial> Plugin for MaterialPlugin<M> {
                 .add_render_command::<AlphaMask3d, DrawMaterial<M>>()
                 .init_resource::<MaterialPipeline<M>>()
                 .init_resource::<SpecializedMeshPipelines<MaterialPipeline<M>>>()
+                .add_system_to_stage(RenderStage::Extract, extract_materials::<M>)
                 .add_system_to_stage(RenderStage::Queue, queue_material_meshes::<M>);
         }
     }
 }
 
+fn extract_materials<M: SpecializedMaterial>(
+    mut commands: Commands,
+    query: Query<(Entity, &ComputedVisibility, &Handle<M>)>,
+    mut prev_len: Local<usize>,
+) {
+    let mut materials = Vec::with_capacity(*prev_len);
+    for (entity, computed_visibility, material) in query.iter() {
+        if computed_visibility.is_visible {
+            materials.push((entity, (material.clone_weak(),)));
+        }
+    }
+    *prev_len = materials.len();
+    commands.insert_or_spawn_batch(materials);
```

Comment:
```suggestion
    commands.insert_or_spawn_batch(query.iter()
             .filter(|(_, visiblility, _)| visibility.is_visible)
             .map(|(entity, _, handle)| (entity, handle.clone_weak()));
```

Try this? It should avoid the Vec allocation and the local.